File: frontend/src/utils/providerDiagnostics.js
function finiteNumber(value, fallback = 0) {
  const number = Number(value);
  return Number.isFinite(number) ? number : fallback;
}

function safeText(value) {
  return String(value || "").trim();
}

function safeEndpoint(scheme, host) {
  const normalizedScheme = safeText(scheme).toLowerCase();
  const normalizedHost = safeText(host);
  if (!normalizedHost || !["http", "https"].includes(normalizedScheme)) return "";
  return `${normalizedScheme}://${normalizedHost}`;
}

export function normalizeProviderDiagnostics(snapshot, now = Date.now()) {
  const source = snapshot && typeof snapshot === "object" ? snapshot : {};
  const generatedAtUnixMs = finiteNumber(source.generatedAtUnixMs, finiteNumber(now, Date.now()));
  const channels = Array.isArray(source.channels)
    ? source.channels.filter((channel) => channel && typeof channel === "object").map((channel, index) => {
      const cooldownUntilUnixMs = finiteNumber(channel.cooldownUntilUnixMs);
      const cooling = safeText(channel.healthState) === "cooldown";
      return {
        channelId: safeText(channel.channelId) || `channel-${index + 1}`,
        displayName: safeText(channel.displayName) || safeText(channel.modelId) || `渠道 ${index + 1}`,
        groupName: safeText(channel.groupName),
        provider: safeText(channel.provider),
        protocolMode: safeText(channel.protocolMode),
        protocolGroup: safeText(channel.protocolGroup),
        modelId: safeText(channel.modelId),
        endpoint: safeEndpoint(channel.endpointScheme, channel.endpointHost),
        contextWindowTokens: Math.max(0, finiteNumber(channel.contextWindowTokens)),
        maxCompletionTokens: Math.max(0, finiteNumber(channel.maxCompletionTokens)),
        credentialConfigured: Boolean(channel.credentialConfigured),
        customHeadersConfigured: Boolean(channel.customHeadersConfigured),
        healthState: cooling ? "cooldown" : "ready",
        cooldownUntilUnixMs: cooling ? cooldownUntilUnixMs : 0,
      };
    })
    : [];
  const cache = source.modelCatalogCache && typeof source.modelCatalogCache === "object"
    ? source.modelCatalogCache
    : {};
  const routerAvailable = Boolean(source.routerAvailable);
  const state = ["ready", "unavailable", "error"].includes(safeText(source.state))
    ? safeText(source.state)
    : (routerAvailable ? "ready" : "unavailable");
  return {
    generatedAtUnixMs,
    state,
    errorCode: safeText(source.errorCode),
    routerAvailable,
    channels,
    readyCount: channels.filter((channel) => channel.healthState === "ready").length,
    cooldownCount: channels.filter((channel) => channel.healthState === "cooldown").length,
    modelCatalogCache: {
      entryCount: Math.max(0, finiteNumber(cache.entryCount)),
      ttlSeconds: Math.max(0, finiteNumber(cache.ttlSeconds)),
      oldestStoredAtUnixMs: Math.max(0, finiteNumber(cache.oldestStoredAtUnixMs)),
      nextExpiryAtUnixMs: Math.max(0, finiteNumber(cache.nextExpiryAtUnixMs)),
    },
  };
}

export function formatProviderCooldown(untilUnixMs, now = Date.now()) {
  const until = finiteNumber(untilUnixMs);
  const remaining = until - finiteNumber(now, Date.now());
  if (remaining <= 0) return "已到恢复时间，请刷新确认";
  if (remaining < 60_000) return "不到 1 分钟后恢复";
  const totalMinutes = Math.ceil(remaining / 60_000);
  if (totalMinutes < 60) return `${totalMinutes} 分钟后恢复`;
  const hours = Math.floor(totalMinutes / 60);
  const minutes = totalMinutes % 60;
  if (minutes > 0) return `${hours} 小时 ${minutes} 分钟后恢复`;
  return `${hours} 小时后恢复`;
}

export function formatDiagnosticTimestamp(unixMs, locale) {
  const value = finiteNumber(unixMs);
  if (value <= 0) return "—";
  return new Intl.DateTimeFormat(locale || undefined, {
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  }).format(new Date(value));
}

File: internal/bridge/provider_diagnostics.go
package bridge

import (
	"time"

	"cursor/internal/client"
)

// ProviderDiagnosticsSnapshot is the Wails-safe provider runtime view.
type ProviderDiagnosticsSnapshot struct {
	GeneratedAtUnixMS int64                       `json:"generatedAtUnixMs"`
	State             string                      `json:"state"`
	ErrorCode         string                      `json:"errorCode,omitempty"`
	RouterAvailable   bool                        `json:"routerAvailable"`
	Channels          []ProviderChannelDiagnostic `json:"channels"`
	ModelCatalogCache ModelCatalogCacheDiagnostic `json:"modelCatalogCache"`
}

type ProviderChannelDiagnostic struct {
	ChannelID               string `json:"channelId"`
	DisplayName             string `json:"displayName"`
	GroupName               string `json:"groupName,omitempty"`
	Provider                string `json:"provider"`
	ProtocolMode            string `json:"protocolMode"`
	ProtocolGroup           string `json:"protocolGroup"`
	ModelID                 string `json:"modelId"`
	EndpointScheme          string `json:"endpointScheme,omitempty"`
	EndpointHost            string `json:"endpointHost,omitempty"`
	ContextWindowTokens     int    `json:"contextWindowTokens,omitempty"`
	MaxCompletionTokens     int    `json:"maxCompletionTokens,omitempty"`
	CredentialConfigured    bool   `json:"credentialConfigured"`
	CustomHeadersConfigured bool   `json:"customHeadersConfigured"`
	HealthState             string `json:"healthState"`
	CooldownUntilUnixMS     int64  `json:"cooldownUntilUnixMs,omitempty"`
}

type ModelCatalogCacheDiagnostic struct {
	EntryCount           int   `json:"entryCount"`
	TTLSeconds           int64 `json:"ttlSeconds"`
	OldestStoredAtUnixMS int64 `json:"oldestStoredAtUnixMs,omitempty"`
	NextExpiryAtUnixMS   int64 `json:"nextExpiryAtUnixMs,omitempty"`
}

// GetProviderDiagnostics performs no upstream request and exposes no raw channel object.
func (s *ProxyService) GetProviderDiagnostics() ProviderDiagnosticsSnapshot {
	if s == nil || s.core == nil {
		return ProviderDiagnosticsSnapshot{
			GeneratedAtUnixMS: time.Now().UnixMilli(),
			State:             "unavailable",
			ErrorCode:         "bridge_service_unavailable",
		}
	}
	return projectProviderDiagnostics(s.core.GetProviderDiagnostics())
}

func projectProviderDiagnostics(source client.ProviderDiagnosticsSnapshot) ProviderDiagnosticsSnapshot {
	result := ProviderDiagnosticsSnapshot{
		GeneratedAtUnixMS: source.GeneratedAtUnixMS,
		State:             source.State,
		ErrorCode:         source.ErrorCode,
		RouterAvailable:   source.RouterAvailable,
		Channels:          make([]ProviderChannelDiagnostic, 0, len(source.Channels)),
		ModelCatalogCache: ModelCatalogCacheDiagnostic{
			EntryCount:           source.ModelCatalogCache.EntryCount,
			TTLSeconds:           source.ModelCatalogCache.TTLSeconds,
			OldestStoredAtUnixMS: source.ModelCatalogCache.OldestStoredAtUnixMS,
			NextExpiryAtUnixMS:   source.ModelCatalogCache.NextExpiryAtUnixMS,
		},
	}
	for _, channel := range source.Channels {
		result.Channels = append(result.Channels, ProviderChannelDiagnostic{
			ChannelID:               channel.ChannelID,
			DisplayName:             channel.DisplayName,
			GroupName:               channel.GroupName,
			Provider:                channel.Provider,
			ProtocolMode:            channel.ProtocolMode,
			ProtocolGroup:           channel.ProtocolGroup,
			ModelID:                 channel.ModelID,
			EndpointScheme:          channel.EndpointScheme,
			EndpointHost:            channel.EndpointHost,
			ContextWindowTokens:     channel.ContextWindowTokens,
			MaxCompletionTokens:     channel.MaxCompletionTokens,
			CredentialConfigured:    channel.CredentialConfigured,
			CustomHeadersConfigured: channel.CustomHeadersConfigured,
			HealthState:             channel.HealthState,
			CooldownUntilUnixMS:     channel.CooldownUntilUnixMS,
		})
	}
	return result
}
